### PR TITLE
Let mlas use session thread pool

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 
1. Let mlas use session thread pool
2. Remove onnxruntime_USE_MLAS cmake option
3. Remove the win32 thread pool code inside mlas

mlas will:
1. use ort thread pool if it get passed in
2. use openmp if the threadpool parameter is nullptr
3. run single threaded if the threadpool parameter is nullptr and openmp is disabled.

Mlas got the threadpool pointer from InferenceSession.

**Motivation and Context**
Let mlas use session thread pool, so that we can fully control the threads.

To avoid deadlock,  for each session, onnxruntime may at most have 3 thread pool instances:
1. session thread pool, for mlas. Controlled by session option
2. parallel executor thread pool
3. lstm thread pool, just for lstm

Before this change, we have 4: the avoid three and a windows thread pool.

### CODE REVIEW

**Comment by pranavsharma** [comment]
Is this a stop gap measure?

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -75,6 +75,7 @@ option(onnxruntime_CROSS_COMPILING "Cross compiling onnx runtime" OFF)
 option(onnxruntime_BUILD_SERVER "Build ONNX Runtime Server" OFF)
 option(onnxruntime_USE_FULL_PROTOBUF "Use full protobuf" OFF)
 option(onnxruntime_DISABLE_CONTRIB_OPS "Disable contrib ops" OFF)
+option(onnxruntime_USE_WIN32_THREADPOOL "Use the default per-process Win32 threadpool" OFF)
 option(onnxruntime_USE_EIGEN_THREADPOOL "Use eigen threadpool. Otherwise OpenMP or a homemade one will be used" OFF)
```
Comment (skottmckay) [inline cmake/CMakeLists.txt]: This description would make me think I should be enabling this. Could it maybe say if/when you'd actually want to enable it and why so it's clear this is probably something you want to leave turned off?

File: include/onnxruntime/core/framework/ort_event.h [hunk 2]
```diff
@@ -11,8 +15,11 @@ struct OnnxRuntimeEvent {
   bool finished = false;
   OnnxRuntimeEvent() = default;
 
- private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(OnnxRuntimeEvent);
 };
 
 using ORT_EVENT = OnnxRuntimeEvent*;
+
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/ort_event.h]: Do we need this 'using' statement? Aliases that hide that something is a pointer often make it harder to read the code.

Comment (snnn) [reply include/onnxruntime/core/framework/ort_event.h]: Will. I'll add another implementation of this ORT_EVENT, that is just a win32 handle.

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc [hunk 3]
```diff
@@ -874,7 +884,7 @@ void UniDirectionalLstm<T>::Compute(const gsl::span<const T>& inputs_arg,
       }
     };
 
-    ExecuteLambdaInParallel("Processing batch", hidden_gemm_and_activations, batch_size_, fused_hidden_rows, ttp_, logger_);
+    ExecuteLambdaInParallel("Processing batch", hidden_gemm_and_activations, batch_size_, fused_hidden_rows, lstm_tp_, logger_);
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc]: If we use the mlas_tp_ here, and pass nullptr in the calls to ComputeGemm on line 836 would that mean we don't need a threadpool in each LSTM node?

File: onnxruntime/core/providers/cpu/math/logsoftmax.cc [hunk 4]
```diff
@@ -12,6 +14,9 @@ namespace onnxruntime {
 
 template <>
 Status LogSoftmax<float>::Compute(OpKernelContext* ctx) const {
+  auto ctx_internal = static_cast<OpKernelContextInternal*>(ctx);
+  auto tp = ctx_internal->GetOperatorThreadPool();
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/math/logsoftmax.cc]: 'auto*' instead of just 'auto'

File: onnxruntime/core/util/math_cpu.cc [hunk 5]
```diff
@@ -62,28 +63,28 @@ EIGEN_MATMUL_FUNCTION(uint64_t)
...
-                              const int64_t N, const int64_t K, float alpha, const float* A, const float* B, float beta,
-                              float* C, CPUMathUtil* /*provider*/) {
+void Gemm<float, ThreadPool>(const CBLAS_TRANSPOSE TransA, const CBLAS_TRANSPOSE TransB, const int64_t M,
+                             const int64_t N, const int64_t K, float alpha, const float* A, const float* B, float beta,
+                             float* C, ThreadPool* threadpool) {
   int lda = static_cast<int>((TransA == CblasNoTrans) ? K : M);
   int ldb = static_cast<int>((TransB == CblasNoTrans) ? N : K);
   // TODO: Make this use the operator threadpool
-  MlasSgemm(TransA, TransB, M, N, K, alpha, A, lda, B, ldb, beta, C, N, nullptr);
+  MlasSgemm(TransA, TransB, M, N, K, alpha, A, lda, B, ldb, beta, C, N, threadpool);
```
Comment (skottmckay) [inline onnxruntime/core/util/math_cpu.cc]: outdated

File: onnxruntime/core/providers/cpu/math/softmax_shared.cc [hunk 6]
```diff
@@ -65,7 +66,7 @@ common::Status SoftmaxCPU(const int64_t N,
   // Put the intermediate result X - max(X) into Y by first copying X to Y, and then subtracting max from each entry
   gsl::copy(gsl::make_span(Xdata, nd), gsl::make_span(Ydata, nd));
 
-  math::Gemm<float, CPUMathUtil>(CblasNoTrans, CblasNoTrans, n, d, 1, -1, rowmax, sum_multiplier, 1, Ydata, nullptr);
+  math::Gemm<float, onnxruntime::concurrency::ThreadPool>(CblasNoTrans, CblasNoTrans, n, d, 1, -1, rowmax, sum_multiplier, 1, Ydata, tp);
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/math/softmax_shared.cc]: Do we need to specify math::Gemm<float, onnxruntime::concurrency::ThreadPool> or can it just be math::Gemm<float> like in gemm.h?

File: onnxruntime/core/providers/cpu/nn/pool_base.h [hunk 7]
```diff
@@ -99,10 +99,13 @@ class LpPool {
 };
 
 class PoolBase {
+ private:
+  static inline bool IsGlobalPooling(const std::string& op_name) {
+    return op_name == "GlobalAveragePool" || op_name == "GlobalMaxPool" || op_name == "GlobalLpPool";
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/nn/pool_base.h]: The 'inline' keyword isn't necessary here as it is implicitly inline

[https://en.cppreference.com/w/cpp/language/inline](https://en.cppreference.com/w/cpp/language/inline)

**Comment by skottmckay** [comment]
> 

These should be private not protected

---
Refers to: onnxruntime/core/session/inference_session.h:417 in 61755a9. [](commit_id = 61755a9c65cc4ef3388c7f299636bf39f9ca9f46, deletion_comment = False)

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc (hunk 3 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc]: Yes. Are you suggesting removing the lstm_tp?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (tracysh) [reply cmake/CMakeLists.txt]: Is this needed at all? I had tried to delete the Win32 thread pool code from MLAS in the first attempt at this.

Comment (snnn) [reply cmake/CMakeLists.txt]: Not necessary.  Let me delete it.

Comment (snnn) [reply cmake/CMakeLists.txt]: @tracysh Do you want mlas be able to run independently without onnxruntime, and the onnxruntime thread pool? If you keep the WIN32_THREADPOOL macro, it means mlas could be still multiple threaded without onnxruntime.

Comment (tracysh) [reply cmake/CMakeLists.txt]: Thanks for checking: in my standalone tests, I just use openmp for threading. The WIN32_THREADPOOL stuff was only brought back for ORT usage.

Comment (snnn) [reply cmake/CMakeLists.txt]: Removed. Please check.

File: include/onnxruntime/core/framework/ort_event.h (hunk 2 above)
Comment (snnn) [reply include/onnxruntime/core/framework/ort_event.h]: Updated.

File: onnxruntime/core/mlas/lib/mlasi.h [hunk 8]
```diff
@@ -423,10 +421,6 @@ struct MLAS_PLATFORM {
     uint32_t NchwcBlockSize;
     uint32_t PreferredBufferAlignment;
 #endif
-
-#if defined(MLAS_USE_WIN32_THREADPOOL)
-    int32_t MaximumThreadCount;
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/mlasi.h]: Can you also remove the now unneeded GetSystemInfo code from platform.cpp?

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc]: Yes. Let mlas provide the threading at the ExecuteLambdaInParallel level, and single thread the GEMM call within the lambda

---
In reply to: [313455794](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r313455794) [](ancestors = 313455794)

File: onnxruntime/core/mlas/lib/mlasi.h (hunk 8 above)
Comment (snnn) [reply onnxruntime/core/mlas/lib/mlasi.h]: Done.

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc (hunk 3 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/rnn/deep_cpu_lstm.cc]: I'd prefer to put it into the next PR.

File: include/onnxruntime/core/framework/ort_event.h (hunk 2 above)
Comment (pranavsharma) [reply include/onnxruntime/core/framework/ort_event.h]: Consider using the Notification class present in the framework folder if it serves the same purpose. It was copied from tensorflow.

File: onnxruntime/core/session/inference_session.h [hunk 9]
```diff
@@ -408,6 +408,9 @@ class InferenceSession {
   ExecutionProviders execution_providers_;
 
  protected:
+  // Threadpool for this session
+  concurrency::ThreadPool* thread_pool_;
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.h]: Would've preferred to use a unique_ptr here so that we don't have to remember to call delete in the destructor. In fact, that's how it was setup before.

Comment (snnn) [reply onnxruntime/core/session/inference_session.h]: I'll make it be able to passed-in from outside, like OrtEnv. Raw pointer is more flexible in that case.

File: include/onnxruntime/core/framework/ort_event.h (hunk 2 above)
Comment (snnn) [reply include/onnxruntime/core/framework/ort_event.h]: Hi @pranavsharma  resolved.

Comment (pranavsharma) [reply include/onnxruntime/core/framework/ort_event.h]: Thanks. Curious, why does windows need a different implementation? Wouldn't Notification work for windows as well?

Comment (snnn) [reply include/onnxruntime/core/framework/ort_event.h]: 1. faster
2. please see the comments in this file.

Comment (snnn) [reply include/onnxruntime/core/framework/ort_event.h]: I'll consider your suggestion.

File: onnxruntime/core/session/inference_session.h (hunk 9 above)
Comment (pranavsharma) [reply onnxruntime/core/session/inference_session.h]: Ok, in that case, please add a comment here saying the same.

File: include/onnxruntime/core/session/environment.h [hunk 10]
```diff
@@ -29,13 +32,20 @@ class Environment {
      Returns whether any runtime environment instance has been initialized.
   */
   static bool IsInitialized() { return is_initialized_; }
+  concurrency::ThreadPool* GetDefaultThreadPool() {
+    std::call_once(tp_once_, [this]() {
+      tp_ = new concurrency::ThreadPool("default", std::max<int>(std::thread::hardware_concurrency() - 1, 1));
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/environment.h]: The previous version used std::thread::hardware_concurrency() / 2. Any reason we changed this? I think @skottmckay had a similar comment.

Comment (pranavsharma) [reply include/onnxruntime/core/session/environment.h]: Also, it looks like the threadpool is always a non-null ptr. Do we need to use a ptr then to pass it around? Can we use a reference instead? A ptr might imply that we need to check for null.

File: onnxruntime/core/session/inference_session.h [hunk 11]
```diff
@@ -135,7 +136,12 @@ class InferenceSession {
     If nullptr, the default LoggingManager MUST have been created previously as it will be used
     for logging. This will use the default logger id in messages.
     See core/common/logging/logging.h for details, and how LoggingManager::DefaultLogger works.
+	@param thread_pool If it's NULL, the session will run as single threaded.
     */
+  explicit InferenceSession(const SessionOptions& session_options,
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.h]: You don't need explicit here.

File: onnxruntime/core/session/inference_session.cc [hunk 12]
```diff
@@ -90,42 +90,49 @@ inline std::basic_string<T> GetCurrentTimeString() {
...
   session_state_.SetProfiler(session_profiler_);
   if (session_options.enable_profiling) {
     StartProfiling(session_options.profile_file_prefix);
   }
 }
 
+InferenceSession::InferenceSession(const SessionOptions& session_options,
+                                   logging::LoggingManager* logging_manager) : InferenceSession(session_options, logging_manager, CreateThreadPool(session_options.session_thread_pool_size)) {
+  //Even it's true, the threadpool could be a NULL pointer
+  own_threadpool_ = true;
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.cc]: can this be part of the initialization list too?

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 13]
```diff
@@ -367,8 +367,18 @@ namespace {
 template <typename Loader>
 OrtStatus* CreateSessionImpl(_In_ const OrtEnv* env, _In_ const OrtSessionOptions* options,
                              Loader loader, _Outptr_ OrtSession** out) {
-  auto sess = std::make_unique<::onnxruntime::InferenceSession>(
-      options == nullptr ? onnxruntime::SessionOptions() : options->value, env->loggingManager);
+  std::unique_ptr<::onnxruntime::InferenceSession> sess;
+  if (options == nullptr || options->value.session_thread_pool_size <= 0) {
```
Comment (pranavsharma) [inline onnxruntime/core/session/onnxruntime_c_api.cc]: I would prefer to be rather explicit about the usage of the env thread pool. Something like a session option that says use_env_threadpool. I don't think this condition really qualifies for the usage of the env thread pool.

File: onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc [hunk 14]
```diff
@@ -45,7 +45,7 @@ UniDirectionalAttnLstm<T>::UniDirectionalAttnLstm(AllocatorPtr allocator,
                                                   const ActivationFuncs::Entry& activation_func_g,
                                                   const ActivationFuncs::Entry& activation_func_h,
                                                   const float clip,
-                                                  onnxruntime::concurrency::ThreadPool& ttp)
+                                                  onnxruntime::concurrency::ThreadPool* ttp)
```
Comment (pranavsharma) [inline onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc]: The PR description said there's going to be a separate lstml thread pool. I don't see it in this iteration. Is that going to be in a separate PR?

File: include/onnxruntime/core/session/environment.h (hunk 10 above)
Comment (pranavsharma) [reply include/onnxruntime/core/session/environment.h]: Should this be called GetThreadPoolFromEnv? reason: the default threadpool is the one that is going to be supplied by the InferenceSession class if there's none provided by the env. The env's threadpool is not really the default one.

File: onnxruntime/core/session/inference_session.cc (hunk 12 above)
Comment (snnn) [reply onnxruntime/core/session/inference_session.cc]: It's hard to make the order correct.

File: include/onnxruntime/core/session/environment.h (hunk 10 above)
Comment (snnn) [reply include/onnxruntime/core/session/environment.h]: I didn't want to change it. @skottmckay suggested me changing it, if I understood correctly.

Comment (snnn) [reply include/onnxruntime/core/session/environment.h]: It could be NULL.

File: onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc (hunk 14 above)
Comment (snnn) [reply onnxruntime/contrib_ops/cpu/attnlstm/uni_dir_attn_lstm.cc]: The separated thread pool is only needed for the UniDirectionalLstm class.

File: onnxruntime/core/session/onnxruntime_c_api.cc (hunk 13 above)
Comment (snnn) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: Even without this PR, they are already sharing threads.  If it worked before, why it doesn't work anymore? Why should I disable it?

File: include/onnxruntime/core/session/environment.h (hunk 10 above)
Comment (snnn) [reply include/onnxruntime/core/session/environment.h]: > Should this be called GetThreadPoolFromEnv

No. This function is in the Env class. I don't think we need to rename it.

Comment (pranavsharma) [reply include/onnxruntime/core/session/environment.h]: > > Should this be called GetThreadPoolFromEnv
> 
> No. This function is in the Env class. I don't think we need to rename it.

The meaning of 'default' is confusing. What does default mean - the one that comes from the env or the one that the session class creates?

File: onnxruntime/core/session/onnxruntime_c_api.cc (hunk 13 above)
Comment (pranavsharma) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: Because now we're being explicit about the existence of a env threadpool.

File: include/onnxruntime/core/session/environment.h (hunk 10 above)
Comment (snnn) [reply include/onnxruntime/core/session/environment.h]: Removed the "default" string.

File: onnxruntime/core/session/onnxruntime_c_api.cc (hunk 13 above)
Comment (snnn) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: But that increases the complexity.  I don't think it's necessary.

File: onnxruntime/core/session/inference_session.h [hunk 15]
```diff
@@ -135,7 +136,12 @@ class InferenceSession {
     If nullptr, the default LoggingManager MUST have been created previously as it will be used
     for logging. This will use the default logger id in messages.
     See core/common/logging/logging.h for details, and how LoggingManager::DefaultLogger works.
+	@param thread_pool If it's NULL, the session will run as single threaded.
     */
+  InferenceSession(const SessionOptions& session_options,
+                   logging::LoggingManager* logging_manager,
+                   concurrency::ThreadPool* thread_pool);
+
   explicit InferenceSession(const SessionOptions& session_options,
```
Comment (linkerzhang) [inline onnxruntime/core/session/inference_session.h]: This may confusing users. 

When writing codes as InferenceSession(session_option), it's using thread pool (not single thread), but when writing codese as InferneceSession(session_option, null, null), it's single thread.

Comment (snnn) [reply onnxruntime/core/session/inference_session.h]: When writing codes as InferenceSession(session_option) and session_options.thread_count=1,it's using thread pool (not single thread). Otherwise it's single threaded.

Comment (snnn) [reply onnxruntime/core/session/inference_session.h]: No matter which function you use,
1. if you pass in a thread pool, the session will use it
2. otherwise, it will create a new one based on session options.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 16]
```diff
@@ -140,12 +144,11 @@ typedef enum OrtErrorCode {
 
 // The actual types defined have an Ort prefix
 ORT_RUNTIME_CLASS(Env);
-ORT_RUNTIME_CLASS(Status);  // nullptr for Status* indicates success
```
Comment (linkerzhang) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: I'd suggest to keep the comments (which is good).

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 17]
```diff
@@ -180,6 +183,9 @@ ORT_API_STATUS(OrtCreateEnvWithCustomLogger, OrtLoggingFunction logging_function
                _In_ const char* logid,
                _Outptr_ OrtEnv** out);
 
+//Don't call OrtReleaseEnv from Dllmain
```
Comment (linkerzhang) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: Put this comment next to "ORT_RUNTIME_CLASS(Env);"?

I'm thinking of avoiding using the macro here to declare all the "OrtRelease***" APIs, which is not that straightforward and clear to users (given we don't have full documentation right now).

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 16 above)
Comment (snnn) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: But for the CreateStatus function, it's not true.

File: onnxruntime/core/session/inference_session.h (hunk 15 above)
Comment (snnn) [reply onnxruntime/core/session/inference_session.h]: I'll revert this change. We can discuss this later.

File: onnxruntime/core/session/inference_session.cc [hunk 18]
```diff
@@ -90,34 +90,33 @@ inline std::basic_string<T> GetCurrentTimeString() {
   OrtStrftime<T>(time_str, sizeof(time_str), GetDateFormatString<T>(), &local_tm);
   return std::basic_string<T>(time_str);
 }
+
+concurrency::ThreadPool* CreateThreadPool(int size) {
+  if (size == 1)
+    return nullptr;  
+  if (size <= 0)
+    size = std::thread::hardware_concurrency() / 2;
+  return new concurrency::ThreadPool("SESSION", size);
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/session/inference_session.cc]: Sorry I wasn't clear enough here. 
We shouldn't do '- 1' here as the current logic that uses the threadpool (e.g. ParallelFor) will only consume threadpool size - 1 threads from the threadpool (+ the main thread), but the logic to determine how many threads are available uses the total threadpool size (MlasGetMaximumThreadCount). 

---
In reply to: [313284294](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r313284294) [](ancestors = 313284294)

[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/core/session/inference_session.cc]: Then we should change mlas.

File: include/onnxruntime/core/session/environment.h [hunk 19]
```diff
@@ -29,13 +32,20 @@ class Environment {
      Returns whether any runtime environment instance has been initialized.
   */
   static bool IsInitialized() { return is_initialized_; }
+  concurrency::ThreadPool* GetThreadPool() {
```
Comment (skottmckay) [inline include/onnxruntime/core/session/environment.h]: Is there going to be a separate PR for this change?

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 16 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: That seems like an edge case though (and hopefully obvious to the caller), so the comment is helpful for every other usage. If you're concerned about OrtCreateStatus maybe note in the comment here or add documentation to OrtCreatetStatus saying it returns a non-nullptr for success.

---
In reply to: [314542648](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r314542648) [](ancestors = 314542648)

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 20]
```diff
@@ -339,7 +342,7 @@ ORT_API_STATUS(OrtGetStringTensorDataLength, _In_ const OrtValue* value, _Out_ s
  * \param value A tensor created from OrtCreateTensor... function.
  * \param s_len total data length, get it from OrtGetStringTensorDataLength
  */
-ORT_API_STATUS(OrtGetStringTensorContent, _In_ const OrtValue* value, _Out_ void* s, size_t s_len,
+ORT_API_STATUS(OrtGetStringTensorContent, _In_ const OrtValue* value, _In_ void* s, size_t s_len,
                _Out_ size_t* offsets, size_t offsets_len);
```
Comment (skottmckay) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: The documentation for this method is not helpful enough. 
Is 's' _In_Out_ as it's not allocated by OrtGetStringTensorContent? 
Should it explicitly say that s_len is the length allocated in 's'?

File: onnxruntime/core/framework/op_kernel_context_internal.h [hunk 21]
```diff
@@ -57,7 +58,8 @@ class OpKernelContextInternal : public OpKernelContext {
 
   const bool& GetTerminateFlag() const noexcept { return terminate_flag_; }
 
-  const onnxruntime::concurrency::ThreadPool* GetOperatorThreadPool() const { return session_state_.GetThreadPool(); }
+  _Ret_maybenull_ const onnxruntime::concurrency::ThreadPool* GetOperatorThreadPool() const { return session_state_.GetThreadPool(); }
+  _Ret_maybenull_ onnxruntime::concurrency::ThreadPool* GetOperatorThreadPool() { return session_state_.GetThreadPool(); }
```
Comment (skottmckay) [inline onnxruntime/core/framework/op_kernel_context_internal.h]: Shouldn't the Session always return a threadpool here (whether that eventually equates to an environment level one or a local one)?

File: onnxruntime/core/framework/session_state.h [hunk 22]
```diff
@@ -175,8 +175,7 @@ class SessionState {
 
   SessionState* GetMutableSubgraphSessionState(onnxruntime::NodeIndex index, const std::string& attribute_name);
 
-  onnxruntime::concurrency::ThreadPool* GetThreadPool() const { return thread_pool_; }
-  void SetThreadPool(onnxruntime::concurrency::ThreadPool* p_pool) { thread_pool_ = p_pool; }
+  concurrency::ThreadPool* GetThreadPool() const { return thread_pool_; }
```
Comment (skottmckay) [inline onnxruntime/core/framework/session_state.h]: A const method returning a non-const internal is generally bad. Please have both methods.

File: onnxruntime/core/framework/session_state.h [hunk 23]
```diff
@@ -232,7 +231,8 @@ class SessionState {
       std::unordered_map<onnxruntime::NodeIndex, std::unordered_map<std::string, std::unique_ptr<SessionState>>>;
   SubgraphSessionStateMap subgraph_session_states_;
 
-  onnxruntime::concurrency::ThreadPool* thread_pool_ = nullptr;
+  //It could be NULL
+  concurrency::ThreadPool* const thread_pool_;
```
Comment (skottmckay) [inline onnxruntime/core/framework/session_state.h]: Should we set it up so that it never returns null, unless there's a reason to allow that?

File: onnxruntime/core/framework/op_kernel_context_internal.h (hunk 21 above)
Comment (skottmckay) [reply onnxruntime/core/framework/op_kernel_context_internal.h]: May need to give this more thought. The optimizer when constant folding creates an OpKernelContext not OpKernelContextInternal, so now any node that needs a threadpool is going to break.

That would suggest we move this method to OpKernelContext (but need to support a null pointer being returned for the threadpool as the optimizer shouldn't need to create a threadpool).

---
In reply to: [314568186](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r314568186) [](ancestors = 314568186)

File: onnxruntime/core/framework/session_state.h (hunk 22 above)
Comment (snnn) [reply onnxruntime/core/framework/session_state.h]: I know your concern. But I'd rather leave the problem to the next guy. Sorry for that. In this special case, it definitely need more discussion.

File: onnxruntime/core/framework/session_state.h (hunk 23 above)
Comment (snnn) [reply onnxruntime/core/framework/session_state.h]: For executing as single threaded.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 20 above)
Comment (snnn) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: It's an error when I reverted/merged the code. I'll drop this change. Thanks for helping me find it.

File: include/onnxruntime/core/session/environment.h (hunk 19 above)
Comment (snnn) [reply include/onnxruntime/core/session/environment.h]: No. I forgot this file. Thanks.

File: onnxruntime/core/framework/op_kernel_context_internal.h (hunk 21 above)
Comment (snnn) [reply onnxruntime/core/framework/op_kernel_context_internal.h]: see: _Ret_maybenull_.

File: onnxruntime/core/framework/session_state.h (hunk 23 above)
Comment (skottmckay) [reply onnxruntime/core/framework/session_state.h]: Seems like there are two choices here:
1) The thread pool implementation handles running single threaded if it's created with zero threads (does what it does currently and executes the provided function on the calling thread).
2) All code that gets a threadpool from session state has to check for nullptr and have one path that uses the threadpool and one path that doesn't

The first seems simpler to me, but either are possible ways to execute single threaded.

---
In reply to: [314569846](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r314569846) [](ancestors = 314569846)

**Comment by pranavsharma** [comment]
LGTM.

**Comment by snnn** [comment]
1. Let mlas use session thread pool
2. Remove onnxruntime_USE_MLAS cmake option
3. Remove the win32 thread pool code inside mlas
mlas will:

1. use ort thread pool if it get passed in
2. use openmp if the threadpool parameter is nullptr
3. run single threaded if the threadpool parameter is nullptr and openmp is disabled.

**Comment by snnn** [comment]
Looks like this change may result crash. Because sometimes the OpKernelContext pointer is not OpKernelContextInternal type.
